types/src/v0/impls/stake_table.rs

```diff
@@ -974,6 +981,137 @@ impl StakeTableFetcher {
         // Process the sorted events and return the resulting stake table.
         validators_from_l1_events(sorted.into_iter().map(|(_, e)| e))
     }
+    /// This function is used to calculate the reward for a block.
+    /// It fetches the initial supply from the token contract.
+    ///
+    /// - We now rely on the `Initialized` event of the token contract (which should only occur once).
+    /// - After locating this event, we fetch its transaction receipt and look for a decoded `Transfer` log
+    /// - If either step fails, the function aborts to prevent incorrect reward calculations.
+    ///
+    /// Relying on mint events directly e.g., searching for mints from the zero address is prone to errors
+    /// because in future when reward withdrawals are supported, there might be more than one mint transfer logs from
+    /// zero address
+    pub async fn fetch_block_reward(&self) -> anyhow::Result<RewardAmount> {
+        let max_events_range = self.l1_client.options().l1_events_max_block_range;
+        let chain_config = *self.chain_config.lock().await;
+
+        let Some(stake_table_contract) = chain_config.stake_table_contract else {
+            bail!("No stake table contract address found in Chain config");
+        };
+
+        let provider = self.l1_client.provider.clone();
+        let stake_table = StakeTableV2::new(stake_table_contract, provider.clone());
+
+        let stake_table_init_block = stake_table
+            .initializedAtBlock()
+            .block(BlockId::finalized())
+            .call()
+            .await?
+            ._0
+            .to::<u64>();
+
+        tracing::info!("stake table init block ={stake_table_init_block}");
+
+        let token_address = stake_table
+            .token()
+            .block(BlockId::finalized())
+            .call()
+            .await
+            .context("Failed to get token address")?
+            ._0;
+
+        let token = EspToken::new(token_address, provider.clone());
+
+        // Attempt to find Initialized event
+        let init_logs = token
+            .Initialized_filter()
+            .from_block(0u64)
+            .to_block(BlockNumberOrTag::Finalized)
+            .query()
+            .await;
+
+        let init_log = if let Ok(init_logs) = init_logs {
+            ensure!(
+                !init_logs.is_empty(),
+                "Token Initialized event logs are empty. This should never happen"
+            );
+
+            let (_, init_log) = init_logs[0].clone();
+
+            tracing::debug!(tx_hash = ?init_log.transaction_hash, "Found token `Initialized` event");
+            init_log
+        } else {
+            const MAX_BLOCKS_SCANNED: u64 = 200_000;
+            let mut total_scanned = 0;
+
+            let mut from_block = stake_table_init_block.saturating_sub(max_events_range);
+            tracing::warn!("Initial token `Initialized` event not found. Scanning backwards from block {from_block}");
+
+            loop {
```

Comment:
This looks correct now. But I think we never execute the scanning code because the first call never fails. How about making a helper function for the scanning code so we can write a unittest for it?